eth/catalyst/tester.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package catalyst
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/core/beacon"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+)
+
+// FullSyncTester is an auxiliary service that allows Geth to perform full sync
+// alone without consensus-layer attached. Users must specify a valid block as
+// the sync target. This tester can be applied to different networks, no matter
+// it's pre-merge or post-merge, but only for full-sync.
+type FullSyncTester struct {
+	api    *ConsensusAPI
+	block  *types.Block
+	closed chan struct{}
+	wg     sync.WaitGroup
+}
+
+// RegisterFullSyncTester registers the full-sync tester service into the node
+// stack for launching and stopping the service controlled by node.
+func RegisterFullSyncTester(stack *node.Node, backend *eth.Ethereum, block *types.Block) (*FullSyncTester, error) {
+	cl := &FullSyncTester{
+		api:    NewConsensusAPI(backend),
+		block:  block,
+		closed: make(chan struct{}),
+	}
+	stack.RegisterLifecycle(cl)
+	return cl, nil
+}
+
+// Start launches the full-sync tester by spinning up a background thread
+// for keeping firing NewPayload-UpdateForkChoice combos with the provided
+// target block, it may or may not trigger the beacon sync depends on if
+// there are protocol peers connected.
+func (tester *FullSyncTester) Start() error {
+	tester.wg.Add(1)
+	go func() {
+		defer tester.wg.Done()
+
+		ticker := time.NewTicker(time.Second * 5)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				// Don't bother downloader in case it's already syncing.
+				if tester.api.eth.Downloader().Synchronising() {
+					continue
+				}
```

Comment:
For some reason, despite this check here, I see this in the logs : 
```
INFO [10-24|11:49:17.672] Syncing beacon headers                   downloaded=512 left=15,816,369 eta=1375h56m34.037s
INFO [10-24|11:49:22.324] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|11:49:25.352] Looking for peers                        peercount=1 tried=30 static=0
INFO [10-24|11:49:25.700] Syncing beacon headers                   downloaded=51712 left=15,765,169 eta=14h15m32.286s
INFO [10-24|11:49:27.324] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|11:49:32.323] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|11:49:33.746] Syncing beacon headers                   downloaded=110,592 left=15,706,289 eta=6h57m35.607s
```

Reply:
Wait, it's snap sync? This thing shouldn't work in snap sync mode at all...

Reply:
Ah, so there is an issue that "reverse-header-sync" is not considered as SYNCING. We should fix it.

Reply:
Yeah, syncmode=full. While it has peers, and actively downloading headers from them, it still spits out this message: 
```
WARN [10-24|12:46:03.514] Post-merge network, but no beacon client seen. Please launch one to follow the chain! 
INFO [10-24|12:46:03.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:08.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:08.881] Looking for peers                        peercount=0 tried=25 static=0
INFO [10-24|12:46:13.521] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:18.452] Syncing beacon headers                   downloaded=512 left=15,816,369 eta=385h33m11.723s
INFO [10-24|12:46:18.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:18.920] Looking for peers                        peercount=1 tried=30 static=0
INFO [10-24|12:46:23.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:26.458] Syncing beacon headers                   downloaded=32768 left=15,784,113 eta=7h4m59.726s
INFO [10-24|12:46:28.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:28.962] Looking for peers                        peercount=2 tried=30 static=0
INFO [10-24|12:46:33.524] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:34.550] Syncing beacon headers                   downloaded=69120 left=15,747,761 eta=3h51m44.555s
INFO [10-24|12:46:38.523] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:39.255] Looking for peers                        peercount=2 tried=37 static=0
INFO [10-24|12:46:42.582] Syncing beacon headers                   downloaded=124,416 left=15,692,465 eta=2h25m10.633s
INFO [10-24|12:46:43.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:48.519] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:50.218] Looking for peers                        peercount=2 tried=40 static=0
INFO [10-24|12:46:50.585] Syncing beacon headers                   downloaded=179,712 left=15,637,169 eta=1h51m45.556s
INFO [10-24|12:46:53.520] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:58.521] Forkchoice requested sync to new head    number=15,816,882 hash=23cd51..dff02b
INFO [10-24|12:46:58.652] Syncing beacon headers                   downloaded=235,008 left=15,581,873 eta=1h34m4.527s

```

Reply:
Yes, it's because the status of downloader is still *NON-SYNCING* but actually it is. So tester keeps firing events non-stop.

Reply:
Yup